### PR TITLE
add example5 - simple stencil2d

### LINKED ISSUE #587: add a stencil2d example to tutorial

### PR DESCRIPTION
**Pull request by haichangsi**

### CODE REVIEW

File: src/example5.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <dr/mhp.hpp>
+#include <fmt/core.h>
+
+namespace mhp = dr::mhp;
+
+using T = uint16_t;
+using MDA = dr::mhp::distributed_mdarray<T, 2>;
+
```
Comment (mateuszpn) [inline src/example5.cpp:12]: Please add short (3 lines or so) description of what the example does (extended description should in README)

File: src/example5.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  mhp::iota(b, 1);
+
+  auto in = dr::mhp::views::submdspan(a.view(), slice_starts, slice_ends);
+  auto out = dr::mhp::views::submdspan(b.view(), slice_starts, slice_ends);
+  auto in_array = &a;
+  auto out_array = &b;
+
+  auto mdspan_stencil_op = [](auto &&v) {
+    auto [in, out] = v;
+    out(0, 0) = (in(-1, 0) + in(0, -1) + in(0, 0) + in(0, 1) + in(1, 0)) / 4;
```
Comment (mateuszpn) [inline src/example5.cpp:33]: Consider changing the element type to floating point - the results of the stencil_op would be probably more clear.

File: src/example5.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  MDA a({arr_size, arr_size}, dist);
+  MDA b({arr_size, arr_size}, dist);
+  mhp::iota(a, 1);
+  mhp::iota(b, 1);
+
+  auto in = dr::mhp::views::submdspan(a.view(), slice_starts, slice_ends);
+  auto out = dr::mhp::views::submdspan(b.view(), slice_starts, slice_ends);
+  auto in_array = &a;
+  auto out_array = &b;
+
```
Comment (mateuszpn) [inline src/example5.cpp:30]: What is the purpose of in_array and out_array?

**Review by mateuszpn — CHANGES_REQUESTED** [review]
Simple run on devcloud resulted in segfault. Why?
```
uab8f7faf7d5cc8a0d0c8bf0d3553a43@idc-beta-batch-head-node:~/work/tutorial-haichangsi$ ./build/src/example5
[1699013467.650602] [idc-beta-batch-head-node:1991980:0]        ib_iface.c:1017 UCX  ERROR ibv_create_cq(cqe=4096) failed: Cannot allocate memory : Please set max locked memory (ulimit -l) to 'unlimited' (current: 4096 kbytes)
Abort(1615247) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
MPIR_Init_thread(176)........:
MPID_Init(1548)..............:
MPIDI_OFI_mpi_init_hook(1632):
create_vni_context(2208).....: OFI endpoint open failed (ofi_init.c:2208:create_vni_context:Input/output error)
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1615247
:
system msg for write_line failure : Bad file descriptor
Abort(1615247) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
MPIR_Init_thread(176)........:
MPID_Init(1548)..............:
MPIDI_OFI_mpi_init_hook(1632):
create_vni_context(2208).....: OFI endpoint open failed (ofi_init.c:2208:create_vni_context:Input/output error)
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1615247
:
system msg for write_line failure : Bad file descriptor
Segmentation fault
```

File: src/example5.cpp (hunk 1 above)
Comment (haichangsi) [reply src/example5.cpp:12]: as I wanted to share the code first I will update README in another PR for all new examples

File: src/example5.cpp (hunk 3 above)
Comment (haichangsi) [reply src/example5.cpp:30]: No purpose, another code version, thank you for noticing ;)

File: src/example5.cpp (hunk 2 above)
Comment (haichangsi) [reply src/example5.cpp:33]: Done

**Comment by haichangsi** [comment]
> Simple run on devcloud resulted in segfault. Why?
> 
> ```
> uab8f7faf7d5cc8a0d0c8bf0d3553a43@idc-beta-batch-head-node:~/work/tutorial-haichangsi$ ./build/src/example5
> [1699013467.650602] [idc-beta-batch-head-node:1991980:0]        ib_iface.c:1017 UCX  ERROR ibv_create_cq(cqe=4096) failed: Cannot allocate memory : Please set max locked memory (ulimit -l) to 'unlimited' (current: 4096 kbytes)
> Abort(1615247) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
> MPIR_Init_thread(176)........:
> MPID_Init(1548)..............:
> MPIDI_OFI_mpi_init_hook(1632):
> create_vni_context(2208).....: OFI endpoint open failed (ofi_init.c:2208:create_vni_context:Input/output error)
> [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1615247
> :
> system msg for write_line failure : Bad file descriptor
> Abort(1615247) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
> MPIR_Init_thread(176)........:
> MPID_Init(1548)..............:
> MPIDI_OFI_mpi_init_hook(1632):
> create_vni_context(2208).....: OFI endpoint open failed (ofi_init.c:2208:create_vni_context:Input/output error)
> [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1615247
> :
> system msg for write_line failure : Bad file descriptor
> Segmentation fault
> ```

as we talked, I think it's because of your specific MPI configuration

File: src/example5.cpp (hunk 1 above)
Comment (mateuszpn) [reply src/example5.cpp:12]: I meant sth very brief, like 
```/* 2d stencil - simple operation on multi-dimensional array */```
or similar

File: src/example5.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#include <dr/mhp.hpp>
+#include <fmt/core.h>
+
+namespace mhp = dr::mhp;
+
+using T = float;
+using MDA = dr::mhp::distributed_mdarray<T, 2>;
+
+int main() {
+#ifdef SYCL_LANGUAGE_VERSION
```
Comment (mateuszpn) [inline src/example5.cpp]: SYCL_LANGUAGE_VERSION is always defined, as we checked the presence of SYCL compiler

**Review by mateuszpn — APPROVED** [review]

File: src/example5.cpp (hunk 4 above)
Comment (haichangsi) [reply src/example5.cpp]: Done

File: src/example5.cpp (hunk 1 above)
Comment (haichangsi) [reply src/example5.cpp:12]: Done

**Pull request merged**